### PR TITLE
added phemex exchange, and upload workspace settings option

### PR DESCRIPTION
**Pull request by lmvdz**
Hi Tucsky

I've created the Phemex exchange files, I've tested with https://thingproxy.freeboard.io/fetch/ because https://cors.aggr.trade/ wasn't working for me...

I've checked to make sure that the price/qty/timestamps are all correct by comparing to their actual trading UI.

https://github.com/phemex/phemex-api-docs/blob/master/Public-Spot-API-en.md#general
https://github.com/phemex/phemex-api-docs/blob/master/Public-Contract-API-en.md#wsapi

image
image

--- edit added another picture with my computer time to show they line up (had to switch to local time in settings)
--- edit I know it says +18 other exchanges on the watermark, but all the other exchanges were disabled in the settings

### CODE REVIEW

**Comment by lmvdz** [comment]
#39 --- possible fix , no way for me to test it. Worked on desktop, uses simple DOM file input click

File: src/worker/exchanges/phemex.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  private channels: { [id: string]: string } = {}
+
+  getUrl() {
+    return 'wss://phemex.com/ws'
+  }
+
+  formatProducts(data) {
+    console.log(data);
```
Comment (Tucsky) [inline src/worker/exchanges/phemex.ts:17]: this should be where we store the contract size for each pairs
image

if format Products return an object, the property of the object are injected in the exchange class,
```
formatProducts(data) {
    const specs = {}
    const products = []

    for (const product of data.data.products) {
      if (product.contractSize) {
        specs[product.symbol] = product.contractSize
      }

      products.push(product.symbol)
    }

    return {
      specs,
      products
    }
  }
```
so then we can use this.specs[* symbol id*] to get the right size of each trade

They seems to have weird price too, scaled price, no idea wtf it is but we could use priceScale / priceRatio from the products endpoint

image

And use it with contractSize

**Comment by Tucsky** [comment]
I've added api.phemex.com to cors.aggr.trade

File: src/worker/exchanges/phemex.ts (hunk 1 above)
Comment (lmvdz) [reply src/worker/exchanges/phemex.ts:17]: Ahh yes thank you for diving deeper into the contract size / price scale.

What I had used was the table values from both contract and spot
image
image
image

Did the math work out to get the correct prices?

Comment (lmvdz) [reply src/worker/exchanges/phemex.ts:17]: cctx has phemex already, so I'm looking into how they are calculating it https://github.com/ccxt/ccxt/blob/b443e0d94ed81df04d76ab7ad3ba93a25859ba64/js/phemex.js#L824

Comment (Tucsky) [reply src/worker/exchanges/phemex.ts:17]: What you did for the price seems to work just fine, I think we don't bother using priceScale and just do the /10000 for perp and /10000 on top of it for spot

Comment (Tucsky) [reply src/worker/exchanges/phemex.ts:17]: image
 for size tho they display it in contract, we display in USD

for eth 1 contract is 0.005 eth so when phemex ui report a trade with 1 size
we need to see about 10$ in aggr (1 * 0.005 * 1985 = 9.925)

Comment (lmvdz) [reply src/worker/exchanges/phemex.ts:17]: It might be better to use the priceScale just incase they decide to change it in the future

Comment (Tucsky) [reply src/worker/exchanges/phemex.ts:17]: Oh for sure, if you figured it out
From all exchanges apis I had to study I never seen anything like this lmao

Comment (lmvdz) [reply src/worker/exchanges/phemex.ts]: There are three cases:

BTCUSD (perp) --- specific case where the contract currency is in USD
SPOT -- Ez calculation
PERP -- all contract currencies are in ETH/SOL/LTC

I pushed the changes to my master

Comment (lmvdz) [reply src/worker/exchanges/phemex.ts]: image

**Comment by lmvdz** [comment]
One thing I am not sure about is that the Phemex exchange doesn't auto connect. Is there another setting somewhere?

**Comment by Tucsky** [comment]
> One thing I am not sure about is that the Phemex exchange doesn't auto connect. Is there another setting somewhere?

It does not because not listed in the default markets to listen to
[src/store/defaultPanes.json](https://github.com/Tucsky/aggr/blob/master/src/store/defaultPanes.json)
it's all here


is Phemex any good ? not sure about adding it by default tbh 
But still nice to have available I guess  😄

**Comment by lmvdz** [comment]
> > One thing I am not sure about is that the Phemex exchange doesn't auto connect. Is there another setting somewhere?
> 
> It does not because not listed in the default markets to listen to
> [src/store/defaultPanes.json](https://github.com/Tucsky/aggr/blob/master/src/store/defaultPanes.json)
> it's all here
> 
> is Phemex any good ? not sure about adding it by default tbh
> But still nice to have available I guess 😄

Sorry, I'm not a good judge of what is good in an exchange...

**Comment by lmvdz** [comment]
Hey Tucsky, I noticed audio wasn't playing on my iPad, so I fixed it in 7aadc5e8878984b236c60d599d7eff8f437c0a8a

**Review by Tucsky — APPROVED** [review]

**Pull request merged**